source/common/http/filter/gzip_filter.h

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include "envoy/http/filter.h"
+#include "envoy/http/header_map.h"
+#include "envoy/json/json_object.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/compressor/zlib_compressor_impl.h"
+#include "common/http/header_map_impl.h"
+#include "common/json/config_schemas.h"
+#include "common/json/json_validator.h"
+
+namespace Envoy {
+namespace Http {
+
+/**
+ * Configuration for the gzip fiter.
+ */
+class GzipFilterConfig : Json::Validator {
+public:
+  GzipFilterConfig(const Json::Object& json_config)
+      : Json::Validator(json_config, Json::Schema::GZIP_HTTP_FILTER_SCHEMA),
+        compression_level_(json_config.getString("compression_level", "default")),
+        compression_strategy_(json_config.getString("compression_strategy", "default")),
+        content_types_(json_config.getStringArray("content_types", true)),
+        memory_level_(json_config.getInteger("memory_level", 8)), window_bits_{31} {}
+
+  uint64_t getMemoryLevel() const { return memory_level_; }
+
+  uint64_t getWindowBits() const { return window_bits_; }
+
+  std::vector<std::string> getContentTypes() const { return content_types_; }
+
+  Compressor::ZlibCompressorImpl::CompressionLevel getCompressionLevel() const {
+    if (compression_level_ == "best") {
+      return Compressor::ZlibCompressorImpl::CompressionLevel::Best;
+    }
+    if (compression_level_ == "speed") {
+      return Compressor::ZlibCompressorImpl::CompressionLevel::Speed;
+    }
+
+    return Compressor::ZlibCompressorImpl::CompressionLevel::Standard;
+  }
+
+  Compressor::ZlibCompressorImpl::CompressionStrategy getCompressionStrategy() const {
+    if (compression_level_ == "rle") {
+      return Compressor::ZlibCompressorImpl::CompressionStrategy::Rle;
+    }
+    if (compression_level_ == "filtered") {
+      return Compressor::ZlibCompressorImpl::CompressionStrategy::Filtered;
+    }
+    if (compression_level_ == "huffman") {
+      return Compressor::ZlibCompressorImpl::CompressionStrategy::Huffman;
+    }
+
+    return Compressor::ZlibCompressorImpl::CompressionStrategy::Standard;
+  }
+
+private:
+  const std::string compression_level_{};
+  const std::string compression_strategy_{};
```

Comment:
It might make more sense to do the string->enum resolution at construction time and then store the enum here instead. You'd still retain the above functions but make them static functions that take the string and return enum, and call them in the ctor initializer list.